File: src/api/gitlab.rs
use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};

pub struct GitLabClient {
    client: Client,
    base_url: String,
    token: String,
}

#[derive(Debug, Serialize)]
struct CreateMergeRequestPayload {
    source_branch: String,
    target_branch: String,
    title: String,
    description: String,
    remove_source_branch: bool,
}

#[derive(Debug, Deserialize)]
struct MergeRequest {
    web_url: String,
}

#[derive(Debug, Deserialize)]
struct Project {
    id: u64,
}

impl GitLabClient {
    pub fn new(base_url: String, token: String) -> Self {
        Self {
            client: Client::new(),
            base_url,
            token,
        }
    }

    pub async fn create_merge_request(
        &self,
        project_path: &str,
        source_branch: &str,
        target_branch: &str,
        title: &str,
        description: &str,
    ) -> Result<String> {
        let project_id = self.get_project_id(project_path).await?;

        let payload = CreateMergeRequestPayload {
            source_branch: source_branch.to_string(),
            target_branch: target_branch.to_string(),
            title: title.to_string(),
            description: description.to_string(),
            remove_source_branch: true,
        };

        let url = format!(
            "{}/api/v4/projects/{}/merge_requests",
            self.base_url, project_id
        );

        let response = self
            .client
            .post(&url)
            .header("PRIVATE-TOKEN", &self.token)
            .json(&payload)
            .send()
            .await
            .context("Failed to send merge request creation request")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("GitLab API error ({}): {}", status, text);
        }

        let mr = response
            .json::<MergeRequest>()
            .await
            .context("Failed to parse merge request response")?;

        Ok(mr.web_url)
    }

    async fn get_project_id(&self, project_path: &str) -> Result<u64> {
        let encoded_path = urlencoding::encode(project_path);
        let url = format!("{}/api/v4/projects/{}", self.base_url, encoded_path);

        let response = self
            .client
            .get(&url)
            .header("PRIVATE-TOKEN", &self.token)
            .send()
            .await
            .context("Failed to fetch project information")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("GitLab API error ({}): {}", status, text);
        }

        let project = response
            .json::<Project>()
            .await
            .context("Failed to parse project response")?;

        Ok(project.id)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gitlab_client_creation() {
        let client = GitLabClient::new(
            "https://git.example.com".to_string(),
            "test-token".to_string(),
        );
        assert_eq!(client.base_url, "https://git.example.com");
        assert_eq!(client.token, "test-token");
    }
}

File: internal/providers/gitlab/gitlab.go
package gitlab

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type Client struct {
	baseURL string
	token   string
	http    *http.Client
}

func NewClient(baseURL, token string) *Client {
	return &Client{
		baseURL: strings.TrimRight(baseURL, "/"),
		token:   token,
		http:    &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *Client) CreateMergeRequest(projectPath, sourceBranch, targetBranch, title, description string) (string, error) {
	projectID, err := c.getProjectID(projectPath)
	if err != nil {
		return "", err
	}

	payload := map[string]any{
		"source_branch":        sourceBranch,
		"target_branch":        targetBranch,
		"title":                title,
		"description":          description,
		"remove_source_branch": true,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/api/v4/projects/%d/merge_requests", c.baseURL, projectID), bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("PRIVATE-TOKEN", c.token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("gitlab api error (%d): %s", resp.StatusCode, string(data))
	}

	var result struct {
		WebURL string `json:"web_url"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return "", err
	}
	return result.WebURL, nil
}

func (c *Client) getProjectID(projectPath string) (int64, error) {
	encoded := url.PathEscape(projectPath)
	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/api/v4/projects/%s", c.baseURL, encoded), nil)
	if err != nil {
		return 0, err
	}
	req.Header.Set("PRIVATE-TOKEN", c.token)

	resp, err := c.http.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return 0, err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return 0, fmt.Errorf("gitlab api error (%d): %s", resp.StatusCode, string(data))
	}

	var result struct {
		ID int64 `json:"id"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return 0, err
	}
	return result.ID, nil
}
